main.cpp

```diff
@@ -24,408 +22,153 @@
     FloatType& subtract(float v);
     FloatType& multiply(float v);
     FloatType& divide(float v);
-    // FloatType args
-    FloatType& add(const FloatType& ft);
-    FloatType& subtract(const FloatType& ft);
-    FloatType& multiply(const FloatType& ft);
-    FloatType& divide(const FloatType& ft);
-    // DoubleType args
-    FloatType& add(const DoubleType& dt);
-    FloatType& subtract(const DoubleType& dt);
-    FloatType& multiply(const DoubleType& dt);
-    FloatType& divide(const DoubleType& dt);
-    // IntType args
-    FloatType& add(const IntType& it);
-    FloatType& subtract(const IntType& it);
-    FloatType& multiply(const IntType& it);
-    FloatType& divide(const IntType& it);
+	
+private:
+	float* m_value { nullptr };
 };
 
 struct DoubleType
 {
-    DoubleType() {}
-    DoubleType(double val) : value(new double(val)) {}
-    ~DoubleType() { delete value; }
+    DoubleType(double val = 0.0) : m_value(new double(val)) {}
+    ~DoubleType()
+	{
+		delete m_value;
+		m_value = nullptr;
+	}
 
-    double* value;
+	operator double() { return *m_value; }
 
-    // primitives
     DoubleType& add(double v);
     DoubleType& subtract(double v);
     DoubleType& multiply(double v);
     DoubleType& divide(double v);
-    // FloatType args
-    DoubleType& add(const FloatType& ft);
-    DoubleType& subtract(const FloatType& ft);
-    DoubleType& multiply(const FloatType& ft);
-    DoubleType& divide(const FloatType& ft);
-    // DoubleType args
-    DoubleType& add(const DoubleType& dt);
-    DoubleType& subtract(const DoubleType& dt);
-    DoubleType& multiply(const DoubleType& dt);
-    DoubleType& divide(const DoubleType& dt);
-    // IntType args
-    DoubleType& add(const IntType& it);
-    DoubleType& subtract(const IntType& it);
-    DoubleType& multiply(const IntType& it);
-    DoubleType& divide(const IntType& it);
+	
+private:
+	double* m_value { nullptr };
 };
 
 struct IntType
 {
-    IntType() {}
-    IntType(int val) : value(new int(val)) {}
-    ~IntType() { delete value; }
-
-    int* value;
+    IntType(int val = 0) : m_value(new int(val)) {}
+    ~IntType()
+	{
+		delete m_value;
+		m_value = nullptr;
+	}
 
-    // primitives
+	operator int() { return *m_value; }
+	
     IntType& add(int v);
     IntType& subtract(int v);
     IntType& multiply(int v);
     IntType& divide(int v);
-    // FloatType args
-    IntType& add(const FloatType& ft);
-    IntType& subtract(const FloatType& ft);
-    IntType& multiply(const FloatType& ft);
-    IntType& divide(const FloatType& ft);
-    // DoubleType args
-    IntType& add(const DoubleType& dt);
-    IntType& subtract(const DoubleType& dt);
-    IntType& multiply(const DoubleType& dt);
-    IntType& divide(const DoubleType& dt);
-    // IntType args
-    IntType& add(const IntType& it);
-    IntType& subtract(const IntType& it);
-    IntType& multiply(const IntType& it);
-    IntType& divide(const IntType& it);
+
+private:
+	int* m_value { nullptr };
```

Comment:
- [ ] FIXME: don't use `m_` in variable names.  Why did you change this?  You're just supposed to make it private, not rename it.